[JENKINS-60848] Improve RSS title wording + Add more RSS tests

jeffret-b opened the pull request:
See [JENKINS-60848](https://issues.jenkins-ci.org/browse/JENKINS-60848)

Improve RSS feed title wording. Make them up-to-date, consistent, and at least somewhat sensible.

Add more tests for RSS as it has been broken several times the past year.

Remove a long unused bit for never implemented trackback.

### Proposed changelog entries

* Improve RSS feed titles.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [n/a] For dependency updates: links to external changelogs and, if possible, full diffs

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a JIRA issue should exist and be labeled as `lts-candidate`

jeffret-b committed:
Add more tests.

Also a little cleanup of RSS.
Fix the broken "All all ..." title and description.
Remove a long unused bit for never implemented trackback.

Wadeck commented on test/src/test/java/hudson/model/RSSTest.java:
```diff
@@ -63,16 +64,89 @@ public void absoluteURLsPresentInRSS_evenWithoutRootUrlSetup() throws Exception
+        Node firstBuild = entries.item(0);
+        assertThat(firstBuild.getChildNodes().getLength(), is(5));
+        assertThat(getSingleNode(firstBuild, "title").getTextContent(), is("test0 #1 (stable)"));
+        assertNotNull(getSingleNode(firstBuild, "published").getTextContent());
```
:x: Does not cover the regression corrected by https://github.com/jenkinsci/jenkins/pull/4148. The "published" was empty and so, is still not null.

jeffret-b replied:
Added checks to validate the dates. There are advantages and disadvantages to canned data. It can mask or complicate other things. There isn't any real need for canned data here.

Wadeck reviewed (changes requested):
I originally proposed to use a "snapshot" approach. This would have covered:
- the missing blank line or whitespace issue that we encountered during the facet development (and was corrected "just in time") _(is not covered in this PR and was not covered during dev)_
- the published date being empty _(is not covered in this PR and was not covered during dev)_
- and potentially other cases

Still pushing for this approach to cover the cases.

----

More about the title / description. Try installing RSS Feed Reader extension on Chrome, it will display you what Jenkins proposed as headers.

<details>

<summary>ScreenshotS</summary>

### Headers proposed by Jenkins with their name
![Screenshot_2020-01-21_114347_001](https://user-images.githubusercontent.com/2662497/72798598-baa3e500-3c43-11ea-9790-6b05d4166191.png)

### RSS feed names used by the extension
![Screenshot_2020-01-21_114528_001](https://user-images.githubusercontent.com/2662497/72798631-c98a9780-3c43-11ea-988a-e1892738aa35.png)

</details>

Side note: even with Locale being fr-FR, the headers does not change, because they are using the "name" and not the "displayName".

What I want to highlight here, is that instead of a quick (dirty?) fix, it's perhaps better to ensure consistency (and do it in a separate ticket?).

Wadeck commented:
Also a note on the removed web method. It's a breaking change, but it seems that the method is doing nothing and there is no OSS plugins using it, so I am fine with its removal.

Alternative approach could be to just let the method there, removing everything inside, putting a `@Deprecated` with a comment saying it will be deleted in some versions. That could be safer but seems just useless in this case.

WDYT @oleg-nenashev

jeffret-b commented:
> Also a note on the removed web method. It's a breaking change, but it seems that the method is doing nothing and there is no OSS plugins using it, so I am fine with its removal.

Since it doesn't do anything and isn't used it's better to just remove it.

jeffret-b commented on core/src/main/java/hudson/model/RSS.java:
```diff
@@ -75,7 +56,14 @@ public static void doTrackback( Object it, StaplerRequest req, StaplerResponse r
+            fixing.append(title.substring(brokenFragment.length()));
+            fixedTitle = fixing.toString();
+        }
+        req.setAttribute("title",fixedTitle);
```
What's your proposal here Wadeck? "All all builds" is pretty bad.

jeffret-b replied:
Here's what it looks like in Thunderbird:
![Screen Shot 2020-01-21 at 2 39 09 PM](https://user-images.githubusercontent.com/37345299/72845483-f5b21280-3c5b-11ea-8318-9cbccbf76eab.png)

It doesn't make much sense to separate this fix out, though changing the title is better and it probably needs a ticket filed.

Wadeck replied:
As the RSS titles are potentially more debatable, I am in favor of separating the addition of test and the change on the title / description.

The proposal would be to just apply the same naming as in the header => `Jenkins:all (all builds)`

jeffret-b committed:
Increase date time checks.

jeffret-b commented:
This adds tests that check a number of additional things beyond what was previously here. Certainly there are more tests that can be added, but this is a decent improvement over what we had. It's clear that there are big advantages to adding tests when we add functionality or fix problems rather than putting it off for some future time where the needed tests may not be as clear.

@Wadeck what other specific checks do you think need to be added? Maybe those should be added in a separate PR.

Wadeck commented:
> what other specific checks do you think need to be added?

The whitespace problem @jvz encountered initially. The objective of the new tests is to at least ensure we are not allowing regression. Perhaps this one was discovered with the pre-existing test, but not sure :) If it's the case, the PR is fine.

Wadeck commented on test/src/test/java/hudson/model/RSSTest.java:
```diff
@@ -135,6 +201,86 @@ public void latestBuilds() throws Exception {
+        String pubDate = getSingleNode(firstBuild, nodeName).getTextContent();
+        assertNotNull(pubDate);
+        DateFormat formatter = new SimpleDateFormat("EEE, dd MMM yyyy HH:mm:ss zzz");
+        Date date = formatter.parse("Sat, 24 Apr 2010 14:01:00 GMT");
```
```suggestion
        Date date = formatter.parse(pubDate);
```

Otherwise the test does not test what you want.

Additionnally, the `parse` will throw an exception if the date is not parseable, so the `assertNotNull` does not provide any value except confusion.

jeffret-b replied:
I tried with and without. It felt kind of weird and more confusing to have it there without any assertion.

Wadeck commented on test/src/test/java/hudson/model/RSSTest.java:
```diff
@@ -135,6 +201,86 @@ public void latestBuilds() throws Exception {
+
+    private void checkRssTimeNode(Node firstBuild, String nodeName) throws ParseException {
+        String pubDate = getSingleNode(firstBuild, nodeName).getTextContent();
+        assertNotNull(pubDate);
```
```suggestion
        assertThat(pubDate, not(emptyString()));
```

`getTextContent` never returns null. If the node does not exist, it's an NPE before the method being called.
That also prevent the next suggestion to not throw an exception if the string is empty.

jeffret-b committed:
Copy and paste error.

Co-Authored-By: Wadeck Follonier <Wadeck@users.noreply.github.com>

jeffret-b committed:
Assert not empty.

Co-Authored-By: Wadeck Follonier <Wadeck@users.noreply.github.com>

jeffret-b commented:
I'm working on adding more tests around the usages of RSS and making the titles consistent. Additional commits to come.

jeffret-b committed:
Add a bunch more tests and clarify title.

Add a number of more tests. Test the various places that RSS is used.
Improve title wording to something consistent and somewhat sensible.

jeffret-b committed:
Merge branch 'rssTestCoverage' of github.com:jeffret-b/jenkins into rssTestCoverage

Wadeck reviewed (approved):
Nice improvement on the code coverage and the consistency with the titles there :)

Wadeck commented:
@jeffret-b You will surely be interested in https://github.com/jenkinsci/jenkins/pull/4411 as it's modifying also the RSS code, to ensure we are not creating conflicts here.

cc @fcojfernandez

oleg-nenashev commented on core/src/main/java/hudson/model/RSS.java:
```diff
@@ -42,25 +41,6 @@
-    /**
-     * Parses trackback ping.
-     */
-    public static void doTrackback( Object it, StaplerRequest req, StaplerResponse rsp ) throws IOException, ServletException {
```
I was unable to find any usages, but it would be great to have some binary compatibility analysis when such changes are submitted. Also not sure why it needs to be renamed

oleg-nenashev reviewed (approved):
I am approving it in general. Two comments"

* It makes sense to replace "Jenkins" in RSS by an Application name. Some vendors like CloudBees change the app name, and I am not sure hardcoded "Jenkins" is what they would like to see in RSS
* Binary compatibility analysis could have been performed by the PR submitter

I plan to merge it tomorrow if no negative feedback

jeffret-b commented:
Yes, it could make sense to use a configurable name, but I consider that outside the scope of this PR. The goal here is to some get decent consistency between the various headers and titles and eliminate ugly phrasing, along with increasing test coverage to avoid regressions.

I did check for usages of the `doTrackback` across the ecosystem and through history. It was never completed and hasn't been used.

oleg-nenashev commented:
Thanks @jeffret-b . Let's just ship it then
